crates/edr_provider/src/data.rs

```diff
@@ -2249,6 +2271,24 @@ impl<LoggerErrorT: Debug, TimerT: Clone + TimeSinceEpoch> ProviderData<LoggerErr
         }
     }
 
+    /// Wrapper over `Blockchain::spec_at_block_number` that handles error
+    /// conversion.
+    fn spec_at_block_number(
+        &self,
+        block_number: u64,
+        block_spec: &BlockSpec,
+    ) -> Result<SpecId, ProviderError<LoggerErrorT>> {
+        self.blockchain
+            .spec_at_block_number(block_number)
+            .map_err(|err| match err {
+                BlockchainError::UnknownBlockNumber => ProviderError::InvalidBlockNumberOrHash {
```

Comment:
Why not construct a `BlockSpec::Number(block_number)`. Is it important to retain what the source block spec is? This should only be able to fail for either `BlockSpec::Number` or `BlockSpec::Eip1898(Eip1898BlockSpec::Number)`.

Reply:
Do you mean to change the `ProviderError::InvalidBlockNumberOrHash` variant? I didn't want to do that, because it's used in other places as well, and I didn't want to add a new variant, because the test expects the exact error message of the `InvalidBlockNumberOrHash` variant. 😅

Reply:
Hey @Wodann I went ahead and merged this, but happy to address in a follow up.

Reply:
Yeah, that's fine!

I meant that you can use the `block_number` parameter of the function to construct a `BlockSpec` inside the function, as it's guaranteed that the only thing that can fail is a `BlockSpec` with a `Number` variant. Tags and hashes would always resolve to existing block numbers before this stage.

Reply:
Ah gotcha now! Yeah so it should only fail if the number came from `BlockSpec::Number`, but the type system doesn't guarantee this, and if there is a logic error somewhere, it's possible that the number was resolved based on a block tag or hash, yet we'd report it as `BlockSpec::Number`, which would be super confusing, so I'd keep rather keep it as is tbh.